File: src/main/java/team/six/youth/guiqiang/_3_28_p24_SwapNodes.java
package team.six.youth.guiqiang;

import team.six.youth.util.ListNode;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class _3_28_p24_SwapNodes {
    public ListNode swapPairs(ListNode head) {
        if(head == null){
            return null;
        }
        if(head.next == null){
            return head;
        }
        ListNode rethead = head.next;
        ListNode wh = head;
        while(wh.next != null){
            ListNode next1 = wh.next;
            ListNode next2 = next1.next;
            next1.next = wh;
            if (next2 == null){
                wh.next = null;
                break;
            }
            ListNode next3 = next2.next;
            if(next3 == null){
                wh.next = next2;
                break;
            }
            wh.next = next3;
            wh = next2;
        }
        return  rethead;   
    }
}

File: src/main/java/team/six/youth/Tesi1a/3_28-P24-SwapNodes.c
/*************************************************************************
    > File Name: 3_28_Solution.c
    > Author: Tesi1a
    > Mail: tangsilian@gmail.com 
    > Created Time: 2018-03-28 22:36
 ************************************************************************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* swapPairs(struct ListNode* head) {

    if(!head||!head->next)
        return head;
    //使用两个指针
    struct ListNode *first=head,*second=head->next;
    int temp;
    //循环至链表末尾
    while(second!=NULL){
        //交换前后链表里的val值
        temp = first->val;
        first->val = second->val;
        second->val = temp;
        if(second->next!=NULL){
            first = second->next;
            second = first->next;
        }else
            break;
    }
    return head;
}
